duplex_http_call.go

```diff
@@ -265,6 +276,19 @@ func (d *duplexHTTPCall) makeRequest() {
 	if d.onRequestSend != nil {
 		d.onRequestSend(d.request)
 	}
+	// If we're sending a unary request, we need to buffer the request body so
+	// that we can send it again if we need to retry.
+	if d.streamType == StreamTypeUnary {
```

Comment:
The exact same error could happen for streaming calls, too. So I don't think we necessarily want to exclude other kinds of RPCs.

Also, I don't think we want to do so much copying. I think we could instead stash entire `*bytes.Buffer` instances and save them (or defer releasing them back to buffer pool) until we decide we've written too much or the operation completes successfully.

Looking through the code, I think a fix to this problem might be rather intrusive, and we'd want to make sure any buffering isn't impacting the performance in a material way. Perhaps it's worth some discussion in the ticket (or a small design doc in Notion)?